pkg/registry/apis/query/parser.go

```diff
@@ -194,16 +194,30 @@ func getTimeRangeForQuery(parentTimerange, queryTimerange *data.TimeRange) data.
 	}
 }
 
-func (p *queryParser) getValidDataSourceRef(ctx context.Context, ds *data.DataSourceRef, id int64) (*data.DataSourceRef, error) {
+func (p *queryParser) getValidDataSourceRef(ctx context.Context, dataQuery data.DataQuery) (*data.DataSourceRef, error) {
+	ds := dataQuery.Datasource
+	id := dataQuery.DatasourceID
+
 	if ds == nil {
 		if id == 0 {
-			return nil, fmt.Errorf("missing datasource reference or id")
+			return nil, NewErrorWithRefID(dataQuery.RefID, fmt.Errorf("missing datasource reference or id"))
 		}
+
 		if p.legacy == nil {
 			return nil, fmt.Errorf("legacy datasource lookup unsupported (id:%d)", id)
 		}
 		return p.legacy.GetDataSourceFromDeprecatedFields(ctx, "", id)
 	}
+
+	if ds.UID[0] == '$' {
```

Comment:
Why are you only checking the first character here? Isn't this too loose?
Can we check the full string exactly matches instead? 

```suggestion
	if ds.UID == '${datasource}' {
```

I feel like this is safer somehow, but I don't have much context here.
Where is the equivalent, old version of the code, so we can compare what was done when this "Permanent migration" (as Kyle called it) was first introduced?

Also, is the `Type == ""` in this case too? Should this code be moved under the `if ds.Type == "" {` check on line 221? Again I think it's worth checking what the logic was on the original code-path, when this migration was introduced.

Reply:
I've seen two values `$datasource` and `${datasource}`. So I went with `$` in this case. :man_shrugging: 

I put the check after the `ds == nil` check and before the `Type == ""` check since the `Type` check path needs a `uid` which we won't have if do this check last.